tidal-link/src/hs/Sound/Tidal/Clock.hs

```diff
@@ -249,8 +222,35 @@ arcStart = fst
 arcEnd :: (Time, Time) -> Time
 arcEnd = snd
 
-timeToCycles' :: ClockConfig -> Link.SessionState -> Link.Micros -> IO Time
-timeToCycles' config ss time = do
+beatToCycles :: ClockConfig -> Double -> Double
+beatToCycles config beat = beat / (coerce $ cBeatsPerCycle config)
+
+cyclesToBeat :: ClockConfig -> Double -> Double
+cyclesToBeat config cyc = cyc * (coerce $ cBeatsPerCycle config)
+
+-- onSingleTick assumes it runs at beat 0.
+-- The best way to achieve that is to use forceBeatAtTime.
+-- But using forceBeatAtTime means we can not commit its session state.
+getZeroedSessionState :: ClockConfig -> ClockRef -> IO Link.SessionState
+getZeroedSessionState config (ClockRef _ abletonLink) = do
+                            ss <- Link.createAndCaptureAppSessionState abletonLink
+                            nowLink <- liftIO $ Link.clock abletonLink
+                            Link.forceBeatAtTime ss 0 (nowLink + processAhead) (cQuantum config)
+                            Link.destroySessionState ss
```

Comment:
It seems unsafe to use ss after it is destroyed. If we want to ensure ss is freed, then I think we need to move flow control into this function. Here we could accept a function (Link.SessionState -> IO a) as argument and return IO a. We would create ss, run the function, then destroy ss, then return.

Reply:
Alternatively, could we move the destroySessionState into a callback function which we include in the return? New return type becomes `IO (Link.SessionState, IO ())`.